clang/lib/CIR/CodeGen/CIRGenTBAA.cpp

```diff
@@ -21,45 +21,156 @@ CIRGenTBAA::CIRGenTBAA(mlir::MLIRContext *mlirContext,
+}
+
+static bool typeHasMayAlias(clang::QualType qty) {
+  // Tagged types have declarations, and therefore may have attributes.
+  if (auto *td = qty->getAsTagDecl())
+    if (td->hasAttr<MayAliasAttr>())
+      return true;
+
+  // Also look for may_alias as a declaration attribute on a typedef.
+  // FIXME: We should follow GCC and model may_alias as a type attribute
+  // rather than as a declaration attribute.
+  while (auto *tt = qty->getAs<TypedefType>()) {
+    if (tt->getDecl()->hasAttr<MayAliasAttr>())
+      return true;
+    qty = tt->desugar();
+  }
+  return false;
+}
+
+/// Check if the given type is a valid base type to be used in access tags.
+static bool isValidBaseType(clang::QualType qty) {
+  if (const clang::RecordType *tty = qty->getAs<clang::RecordType>()) {
+    const clang::RecordDecl *rd = tty->getDecl()->getDefinition();
+    // Incomplete types are not valid base access types.
+    if (!rd)
+      return false;
+    if (rd->hasFlexibleArrayMember())
+      return false;
+    // rd can be struct, union, class, interface or enum.
+    // For now, we only handle struct and class.
+    if (rd->isStruct() || rd->isClass())
+      return true;
+  }
+  return false;
+}
+
 cir::TBAAAttr CIRGenTBAA::getTypeInfo(clang::QualType qty) {
-  return tbaa_NYI(mlirContext);
+  // At -O0 or relaxed aliasing, TBAA is not emitted for regular types.
+  if (codeGenOpts.OptimizationLevel == 0 || codeGenOpts.RelaxedAliasing) {
+    return nullptr;
+  }
+
+  // If the type has the may_alias attribute (even on a typedef), it is
+  // effectively in the general char alias class.
+  if (typeHasMayAlias(qty)) {
+    llvm_unreachable("NYI");
+    return getChar();
+  }
+  // We need this function to not fall back to returning the "omnipotent char"
+  // type node for aggregate and union types. Otherwise, any dereference of an
+  // aggregate will result into the may-alias access descriptor, meaning all
+  // subsequent accesses to direct and indirect members of that aggregate will
+  // be considered may-alias too.
+  // function.
+  if (isValidBaseType(qty)) {
+    // TODO(cir): support TBAA with struct
+    return tbaa_NYI(mlirContext);
+  }
+
+  const clang::Type *ty = astContext.getCanonicalType(qty).getTypePtr();
+  if (metadataCache.contains(ty)) {
+    return metadataCache[ty];
+  }
+
+  // Note that the following helper call is allowed to add new nodes to the
+  // cache, which invalidates all its previously obtained iterators. So we
+  // first generate the node for the type and then add that node to the
+  // cache.
+  auto typeNode = cir::TBAAScalarTypeDescriptorAttr::get(
+      mlirContext, types.ConvertType(qty));
+  return metadataCache[ty] = typeNode;
 }
 
 TBAAAccessInfo CIRGenTBAA::getAccessInfo(clang::QualType accessType) {
-  return TBAAAccessInfo();
+  // Pointee values may have incomplete types, but they shall never be
+  // dereferenced.
+  if (accessType->isIncompleteType()) {
+    llvm_unreachable("NYI");
```

Comment:
Our internal builds use `-O2` to build our apps, and seems like this path and some of the others will always run under `-O2`. For paths that will unconditionally run, you are better off with, e.g. `assert(!MissingFeature::tbaaIncompleteType())` or something. Can you confirm that all added unrecheables are never executed just yet? An alternative is to add a temporary cc1 flag and only enable tbaa when the flag is used, once you are done with the full work you would remove the flag.

Reply:
> Our internal builds use -O2 to build our apps, and seems like this path and some of the others will always run under -O2

did not encounter the NYI issue when run with `-O2`. it may be related to the struct type.

I think adding `MissingFeatures::tbaaIncompleteType()` would be a good idea.